### ISSUE TITLE
[Proposal: creating disaggregator class]

### ISSUE THREAD

**Issue by nipunbatra** [nilmtk/nilmtk#755]
### Goal: Make it easy to write disaggregation algorithms in NILMTK

Currently, it is very hard for disaggregation algorithm writers to write new methods. They need to know a lot of NILMTK. I am proposing a simple disaggregator interface where they need to know only the PyData stack.

I am proposing that each disaggregator define the following methods:

```python
def partial_fit(train_main, [train_appliance_1, train_appliance_i, ..., train_appliance_n]):
     """
     train_main: pd.DataFrame with pd.DatetimeIndex as index and 1 or more power columns
     train_appliance_i: pd.DataFrame with the same pd.DatetimeIndex as index as train_main and the same 1 or more power columns as train_main
     """
```

An important thing to note for this `partial_fit` is that when called repeatedly over new data it keeps on fitting the new data and improving the model.

Some methods like CO in current implementation can not update the model (atleast in current implementation). For such methods, we could allow `partial_fit` to break when called for the second time. 

Testing could be:

```python
def predict(test_main):
    """
    test_main: pd.DataFrame with pd.DatetimeIndex as index and 1 or more power columns
    returns: [train_appliance_1, train_appliance_i, ..., train_appliance_n]
    """


The methods proposed in #753 would invoke the appropriate train and test methods.

The underlying goal of this proposed implementation is that disaggregation algorithm writers should have an easy time writing algorithms in NILMTK and should know only Pandas, Numpy, Sklearn, and not need to know the intrinsics of NILMTK.

Thoughts @PMeira, @OdysseasKr, @MingjunZhong ?

**Comment by OdysseasKr** [comment]
This sounds like a very flexible design overall. However is actually limiting for more complex algorithms such as NNs. 

For the implementation of [neural-disaggregator](https://github.com/OdysseasKr/neural-disaggregator), due to memory constraints, it was necessary for every epoch to:
- load parts of the data from multiple datasources/timespans from disk into memory
- preprocess (sliding windows, normalization etc)
- shuffle

Using this architecture means that all of the data needs to fit into memory which I would say is not very realistic for NNs. I would like to know if @MingjunZhong 's implementations have similar requirements.

The only way I could see this working with more demanding disaggregators would require the algorithm writer to also write their own data loader.

**Comment by nipunbatra** [comment]
I think that the preprocessing (shuffling included) could also be an argument in the configuration I had mentioned in: #753 I think Jack always intended building pipelines in NILMTK and I think we'd be doing something similar to support neural methods you mention.

Since you mention your implementation does not perform partial fit and you load parts of data from multiple datasources/time-stamps, I would suppose that your algorithms feed upon chunks of dataframe like objects, all of which are required to be in-memory together?

I think the above should be possible by adding some preprocessing function calls to the configuration so that data is loaded appropriately and fed into the `fit` method. How does that sound like?

**Comment by MingjunZhong** [comment]
@OdysseasKr The way that we were following is that 1) you can read all the training data into the memory, which is basically two columns where one is the mains and the other is the appliance; 2) in order to using mini-batches of windows, you only shuffle the labels (or time stamps) so that the code can read the batch data using these labels. That means you does not have to get all the possible window for training which needs a lot of memories. Is that what you are asking about?
@nipunbatra Particularly for seq2point and seq2seq meodels, that means that shuffling is not necessarily included in the preprocessing process. For this class of NN approaches, as the preprocessing include removing means and dividing STD, we could apply them to the data before making mini-batches as described.
So these two points inform that we could allow it to perform on-line training - when new data is read into the memory, they could be treated as new batches of windows.
Why it is called `partial_fit()'? As a user, I could be confused by this name as it is not obvious it is used for training a model.

**Comment by OdysseasKr** [comment]
@nipunbatra OK, this sounds very good! I was not expecting the scope of NILMTK to also include very specific preprocessing needs.

> Since you mention your implementation does not perform partial fit and you load parts of data from multiple datasources/time-stamps, I would suppose that your algorithms feed upon chunks of dataframe like objects, all of which are required to be in-memory together?

It was impossible for all chunks to be in memory. So what I implemented was loading parts of the raw mains and appliance data, preprocessing them accordingly and finally feeding the chunk to the network. This was done multiple times until all of the data were used.

@MingjunZhong Yes this is what I was asking, thanks!

I am thinking that it is called ```partial_fit``` in compliance to the sklearn interface. I don't know if this is ideal for this case since there is no ```fit``` method but I still find it proper naming.

**Comment by MingjunZhong** [comment]
@OdysseasKr I understand the partial_fit and fit methods now. We could include both methods although we are likely to implement partial_fit. So it makes sense to load parts of the raw data from source files. Is it time consuming to repeatedly load a lot of raw data from source files because it requires many epochs?

**Comment by nipunbatra** [comment]
>It was impossible for all chunks to be in memory. So what I implemented was loading parts of the raw mains and appliance data, preprocessing them accordingly and finally feeding the chunk to the network. This was done multiple times until all of the data were used.

This is exactly what I'm suggesting too with the configuration:

For example: the above configuration (after adding preprocessing):
```python
clf = configuration['classifier']
for dataset_name, dataset in configuration['datasets']:
    ds = load_dataset(dataset_name):
        for building_num in dataset['buildings']:
            start, end = dataset['buildings]['start_time'], dataset['buildings]['end_time']
            for appliance in configuration['appliances']:
                for appliance_chunk in load(appliance, configuration['sample_rate'],  start, end):
                     processed_chunk = configuration['preprocess'](appliance_chunk)
                     clf.partial_fit(processed_chunk)
```
 
or, if we can preprocess and reduce the data and then process it once through the network:

```python
clf = configuration['classifier']
processed_chunk_appliance = {}
for dataset_name, dataset in configuration['datasets']:
    ds = load_dataset(dataset_name):
        for building_num in dataset['buildings']:
            start, end = dataset['buildings]['start_time'], dataset['buildings]['end_time']
            for appliance in configuration['appliances']:
            processed_chunk_appliance[appliance]  = []
                for appliance_chunk in load(appliance, configuration['sample_rate'],  start, end):
                     processed_chunk_appliance[appliance].append(configuration['preprocess'](appliance_chunk))
for appliance in configuration['appliances']:
    # can add shuffling here if you want
    clf.fit(processed_chunk_appliance[appliance])
```

In either case, the `clf.fit()` or `clf.partial_fit` is a simple function only requiring us to pass Pandas dataframes.

How does it sound now?

**Comment by MingjunZhong** [comment]
Just for a question, the preprocessing is for an appliance chunk and I assume the preprocessing includes normalisation and as well as resampling which would be necessary as it is very often the data was not equally sampled and also we want to try different sampling rates to test our algorithms. The questions is that we need more than one epochs and so the current configuration will redo the preprocessing many times which is a waste of time. An option would be to do the preprocessing in an independent method and then apply the current one to the preprocessed data.

**Comment by nipunbatra** [comment]
@MingjunZhong @OdysseasKr 
In light of your above comments, how about a preprocessing stage that takes as input the dataset, buildings in each and start and end times in each, and appliance list, along with a preprocessing function and outputs a `processed.h5` binary, which can then be directly passed to `fit`?

I think the key would be to output `processed.h5` in the same manner as the original dataset.h5 format? So that the method that does not require preprocessing works directly?

**Comment by OdysseasKr** [comment]
@nipunbatra The above code looks very clean and seems like a good solution. My main concern here is the ```configuration['preprocess'](appliance_chunk)``` function is not as simple as it seems. Preprocessing can get a very particular process very quickly as new methods are created. Especially when we have sliding windows on input and output. Maybe that's not in the scope of NILMTK

I think that a general preprocessing step is welcome but it would be nice for the algorithm writer to be able to easily overload this function if needed.

> The questions is that we need more than one epochs and so the current configuration will redo the preprocessing many times which is a waste of time.

Agreed. This is a very realistic issue.

> In light of your above comments, how about a preprocessing stage that takes as input the dataset, buildings in each and start and end times in each, and appliance list, along with a preprocessing function and outputs a processed.h5 binary, which can then be directly passed to fit?

This sounds more feasible to me. Essentially the NILMTK plumbing method acts a method that mixes/prepares the data and packages them in a standard way that the classifier can use.

**Comment by MingjunZhong** [comment]
That .h5 file looks fine to me, but as a user I am curious about what was wrapped in that file and as well as the original files so that I want to visualize them. For example, are the data very noisy? are there many missing values? are the data wrongly labeled? I understand that we assume the data we are going to use for training are well preprocessed, but the users do need to visualize them. I think we need an interface for visualizing the data allowing to zoom in and out for example. This interface should be independent to applying the algorithms to the data. This is out of the scope of this proposal; it may need a separate ticket in the future.

**Comment by nipunbatra** [comment]
I think if the .h5 is in the same format as the original datasets, then, it should be trivial for users to visualise the data. This is because NILMTK already provides good visualisation on dataset .h5 files. As a last resort, user can easily pull the data from .h5 and plot.

**Comment by OdysseasKr** [comment]
> I think if the .h5 is in the same format as the original datasets, then, it should be trivial for users to visualise the data. This is because NILMTK already provides good visualisation on dataset .h5 files. As a last resort, user can easily pull the data from .h5 and plot.

I agree with this. Also, if the algorithms are to be used as a black box, I think it's outside of the scope whether users try to plot preprocessed data.

**Comment by MingjunZhong** [comment]
That's great. We can then focus on the algorithms.

**Comment by nipunbatra** [comment]
Also, please see: https://github.com/nilmtk/nilmtk/tree/master/nilmtk/preprocessing
and https://github.com/nilmtk/nilmtk/tree/master/nilmtk/stats
